File: STUDY/SOLUCIONES/PYTHON/04_Funciones/ejercicios_intermedio_solucion.py
"""
Solucion intermedio - Funciones
"""

def apply(fn, items):
    """Aplica fn a cada item."""
    return [fn(x) for x in items]


def compose(f, g):
    """Devuelve f(g(x))."""
    def h(x):
        return f(g(x))
    return h


def filter_by(fn, items):
    """Filtra items segun fn."""
    return [x for x in items if fn(x)]


def main():
    print("apply:", apply(lambda x: x + 1, [1, 2, 3]))
    doble_mas_1 = compose(lambda x: x + 1, lambda x: x * 2)
    print("compose:", doble_mas_1(3) if doble_mas_1 else None)
    print("filter:", filter_by(lambda x: x % 2 == 0, [1, 2, 3, 4]))


if __name__ == "__main__":
    main()

File: STUDY/SOLUCIONES/RUST/05_Traits_Generics/ejercicios_intermedio_solucion.rs
// ============================================
// SOLUCION INTERMEDIO - TRAITS Y GENERICS
// ============================================

struct Pila<T> {
    items: Vec<T>,
}

impl<T> Pila<T> {
    fn new() -> Self {
        Pila { items: Vec::new() }
    }

    fn push(&mut self, item: T) {
        self.items.push(item);
    }

    fn pop(&mut self) -> Option<T> {
        self.items.pop()
    }
}

fn main() {
    let mut p = Pila::new();
    p.push(1);
    println!("pop: {:?}", p.pop());
}
